### ISSUE TITLE
Manual edit prevents next patch

### ISSUE THREAD

**Issue by sta-szek** [helm/helm#10858]
Output of `helm version`: 

`version.BuildInfo{Version:"v3.5.3", GitCommit:"041ce5a2c17a58be0fcd5f5e16fb3e7e95fea622", GitTreeState:"dirty", GoVersion:"go1.15.8"}`

Output of `kubectl version`:  

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.12", GitCommit:"2cb4598553d0a2ca5f3eb7980c99cd8236cc5011", GitTreeState:"clean", BuildDate:"2021-06-16T13:00:29Z", GoVersion:"go1.15.13", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.10", GitCommit:"98d5dc5d36d34a7ee13368a7893dcb400ec4e566", GitTreeState:"clean", BuildDate:"2021-04-15T03:20:25Z", GoVersion:"go1.15.10", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  Kops (own instances somehere, i dont even know)

The issue occurs when you do manual edit, e.g. to `ingress` then the next deployment will not update ingress according to helm chart but it will leave it as is.
That is causing significant issues for us, because sometimes we must do manual fix in production, and next deployment can lead to outage, because Helm does not re-apply itself.

Reproduction steps:

1. `kubectl get ingress cloud-poc -o yaml`
2. add annotation manually:  `nginx.ingress.kubernetes.io/rewrite-target: "/$1"`
3. `kubectl get ingress cloud-poc -o yaml`
4. redeploy - with helm chart that do not have manually added annotation
5. `kubectl get ingress cloud-poc -o yaml`

The expectation is that annotation `nginx.ingress.kubernetes.io/rewrite-target: "/$1"` will be removed after step 4, because the original helm chart does not contain it.

The resource rendered before deployment:

```
upgrade.go:150: [debug] performing update for cloud-poc
# Source: cloud-poc/templates/ingress-external.yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: cloud-poc
  labels:
    app:  cloud-poc
    chart:  cloud-poc-2.4.0-b5a90c9e
    release: cloud-poc
    heritage: Helm
    version: 2.4.0-b5a90c9e
    maintainer_team: XXXX
    spring-boot: "true"
    sidecar.istio.io/inject: "true"
  annotations:
    cert-manager.io/cluster-issuer: XXX
spec:
  rules:
    - host: XXX.net
      http:
        paths:
          - path: /cloud-poc(?!(/(cloud-poc/)*(actuator))).*
            pathType: ImplementationSpecific
            backend:
              service:
                name: istio-ingressgateway
                port:
                  number: 80
```

as you can see, there is no annotation `nginx...` but after deployment it is still preserved.

what i notice is that after step 2 we can see additional "entry" in ingress itself:

```
  - apiVersion: extensions/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:nginx.ingress.kubernetes.io/rewrite-target: {}
    manager: kubectl-edit
    operation: Update
    time: "2022-04-13T09:14:27Z"
```

in the section `managedFields` when i do `kubectl get ingress cloud-poc -o yaml` 

1. is that desired behaviour? why?
2. how we can make Helm to deploy actual Helm chart and do not skip anything? BTW in logs i see: `client.go:510: [debug] Patch Ingress "cloud-poc" in namespace default`

**Comment by alexivanoff** [comment]
as far as I know helm compare resources using previously generated secrets
looks like you deployed the same resource with the same specs and annotations
so helm skipped it

**Comment by sta-szek** [comment]
> as far as I know helm compare resources using previously generated secrets looks like you deployed the same resource with the same specs and annotations so helm skipped it

hmm, that could explain the behaviour: i did not change the ingress.yaml, so it had exactly the same hash.

but then if that is the answer, i have next questions:
1. How to change the behaviour so helm compares helm chart with actual deployment, not with previously deployed helm chart?
2. why we see `client.go:510: [debug] Patch Ingress "cloud-poc" in namespace default` if there was no actual patch?
3.  if there is no solution, would be nice to add flag `--compare-with-k8s` or smth, what do you think?

**Comment by sta-szek** [comment]
what i found so far is that we can do deployment with `--force` but that will lead to another issue -> it will always force deployment even if there is no actual change :) 
and that is also smth we want to avoid -> we often rerun deplyments and we relay on behaviour "no changes, no deployment" because we just want to make our CD deployment green. then pipeline continues

if we do `--force` then we re-deploy on each re-run which will be problematic for us.

i really need smth like `--compare-with-k8s-resources` or `--force-if-manual-changes` - you know what i mean?

**Comment by marckhouzam** [comment]
I believe you are affected by Helm v3's three-way merge.  This is a feature that aims to do exactly what you are seeing: allow users to modify a deployment and then persist that change.  Here are some details: https://helm.sh/docs/faq/changes_since_helm2/#improved-upgrade-strategy-3-way-strategic-merge-patches

As you found, I do believe you can use the `--force` flag to disable the three-way merge. 

> what i found so far is that we can do deployment with --force but that will lead to another issue -> it will always force deployment even if there is no actual change :)

I'm not sure what you mean.  If you do a helm upgrade, you should see a new version of the release even if the yaml file have not changed.  It is kubernetes that will not perform any changes when it sees the yaml has not changed.

**Comment by bacongobbler** [comment]
Closing this issue as answered. Please feel free to comment here and let us know if you have further questions. Thanks!
